src/journal/journald-context.c

```diff
@@ -500,6 +513,70 @@ static int client_context_read_log_ratelimit_burst(ClientContext *c) {
         return safe_atou(value, &c->log_ratelimit_burst);
 }
 
+static int client_context_read_log_regex(
+                ClientContext *c,
+                const char *key,
+                char **regex,
+                pcre2_code **compiled_regex) {
+#if HAVE_PCRE2
+        int r;
+        char *new_regex = NULL;
+
+        assert(c);
+
+        if (!c->cgroup)
+                return 0;
+
+        r = cg_get_xattr_malloc(SYSTEMD_CGROUP_CONTROLLER, c->cgroup, key, &new_regex);
+        if (r == -ENODATA) {
+                /* If attribute is not set, we need to ensure ClientContext does
+                 * not store any old value. */
+                sym_pcre2_code_freep(compiled_regex);
+                *regex = mfree(*regex);
+                return 0;
+        } else if (r < 0) {
+                log_debug_errno(r, "Failed to get %s xattr for %s: %m", key, c->cgroup);
+                return r;
+        }
+
+        /* If regex hasn't changed, then there is no need to compile it again */
+        if (*regex && strcmp(*regex, new_regex) != 0) {
+                new_regex = mfree(new_regex);
+                return 0;
+        }
+
+        sym_pcre2_code_freep(compiled_regex);
```

Comment:
Similarly to `free_and_replace()`, I'd add `pcre2_code_free_and_replace()` to pcre2-util.h and then use it here to replace `compiled_regex` when needed at the end of the function. Until then the new compiled regex should be stored in a local variable to avoid modifying the return variable on error.